include/ecs/ComponentPool.hpp

```diff
@@ -0,0 +1,374 @@
+#include "ComponentPool.h"
+#include <utils/Assert.h>
+
+#define TEMPLATE template<class C, class E>
+#define POOL ComponentPool<C, E>
+#define TRAITS _details::ComponentTraits<C, E>
+
+// https://miro.com/app/board/uXjVK4gF1DI=/?share_link_id=296698570044
+// ^ picture explanations
+
+namespace arch::ecs {
+
+TEMPLATE
+POOL::~ComponentPool() noexcept {
+	if constexpr (not Traits::flag) {
+		for (size_t i = 0; i != _components.size(); ++i) {
+			Traits::deletePage(_components.data(), i, _dense);
+		}
+	}
+}
+
+TEMPLATE
+E& POOL::_sparseAssure(const IdT id) noexcept {
+	const size_t sparsePageNum = qdiv<ETraits::pageSize>(id);
+	// virtually equal to _sparse.resize(sparsePageNum + 1), but capacity >= sparsePageNum + 1
+	while (_sparse.size() != sparsePageNum + 1) {
+		_sparse.emplace_back();
+	}
+
+	auto&& sparsePagePtr = _sparse[sparsePageNum];
```

Comment:
some assertion for out of bounds?

Reply:
while 3 lines before would make it obsolete